залил 3 домашку

smidy2002 opened the pull request.

smidy2002 committed:
залил 3 домашку

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+    long int ans = 0;
+    while (i < n.size())
+    {
+        const string maxValue = "2147483647";
```
std::numeric_limits<uint32_t>.max()

DimohaZadira replied:
то есть строка, построенная по этому значению. Magic Value быть не должно

smidy2002 replied:
заменил "2147483647" на std::numeric_limits<uint32_t>.max()

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+#include <string>
+#include <cmath>
+using namespace std;
+int valid(string n)
```
int -> uint32_t, потому что инт платформозависимый, это кринж
ну и переименую переменную, n это кринж, вообще обычно n натуральное, а тут строка. Так и назови, string_num
Что происходит со строкой, когда ты её передаёшь в функцию по значению?

smidy2002 replied:
не понял вопрос

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+int valid(string n)
+{
+    int i = 0;
+    long int ans = 0;
```
uint64_t

smidy2002 replied:
заменил все int на uint32_t, long был рудиментом

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+    while (i < n.size())
+    {
+        const string maxValue = "2147483647";
+        if(n[0] == 48) { return -1; }
```
Это точно нужно на каждой итерации цикла проверять?

smidy2002 replied:
вынес ифы из тела цикла

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+    {
+        const string maxValue = "2147483647";
+        if(n[0] == 48) { return -1; }
+        if (n[i] > 57 || n[i] < 48) { return -1; }
```
Что такое 57 и 48? Сделай через '0' и '9' сравнения) Я мог и не знать, что '0' это 48. Всегда делай код более читаемым

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+        const string maxValue = "2147483647";
+        if(n[0] == 48) { return -1; }
+        if (n[i] > 57 || n[i] < 48) { return -1; }
+        if (!(n.size() < maxValue.size() || n.size() == maxValue.size() && n <= maxValue))//не смог придумать сам взял у умных дядек
```
Во-первых, разберись, что каждое сравнение делает по отдельности. Во-вторых, расставь скобки и расскажи в ответе к этому комменту, что тут на самом деле происходит :)
1. На какой вопрос семантически отвечает этот иф?
2. Что нужно сделать (человеческими словами), чтобы ответить на вопрос, сформулированный в п.1?
3. Что делает каждая часть этого ифа?

smidy2002 replied:
1. должен отвечать на вопрос влазит ли строка "n" в 32 бита чтобы мы могли ее потом превратить в инт
2. проверит что количество символов менньше чем у максимального инта, или символов столько же и тогда нужно поразрядно проверить что наша стока меньше макс. инта
3. 
(n.size() < maxValue.size() проверяет что у нашей строки меньше разрядов чем у макс инта
n.size() == maxValue.size() проверяет что разрядов столько же сколько у инта 
n <= maxValue если предидущее верно то мы можем посимвольно сравнить что больше, наша строка или макс инт

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+        {
+            return -1;
+        }
+        ans += int(n[i] - 48) * pow(10, n.size() - i - 1);
```
48 -> '0', int -> uint32_t. Надеюсь, ты понимаешь, что происходит в этой строчке)

DimohaZadira replied:
Тут ты столкнёшься с проблемой, а как быть, потому что щас у тебя возвращаемое значение "-1" значит, что что-то пошло не так. Ответ -- функция должна возвращать только 0, если всё ОК, и -1, если что-то пошло не так, а значение меня по указателю. То есть передай в функцию указатель на uint32_t, заведённый в мейне, и меняй в функции значение по этому указателю.

smidy2002 replied:
превращаем строку ипа "abc" в число вида 100a+10b+c

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+    while (n != "5312")
+    {
+        cin >> n;
+        if (valid(n) != -1) { ans += valid(n); }
```
Тут ты дважды вызываешь функцию, это отстой. Вызови один раз, сохрани значение и обработай потом.

DimohaZadira replied:
Тут ты столкнёшься с проблемой, а как быть, потому что щас у тебя возвращаемое значение "-1" значит, что что-то пошло не так. Ответ -- функция должна возвращать только 0, если всё ОК, и -1, если что-то пошло не так, а значение меня по указателю. То есть передай в функцию указатель на uint32_t, заведённый в мейне, и меняй в функции значение по этому указателю.

smidy2002 replied:
![image](https://github.com/user-attachments/assets/ed0c1f23-42e1-44fc-924b-b9b60622cd92)
вот так вроде работает

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+        if (valid(n) != -1) { ans += valid(n); }
+        else
+        {
+            cout << "invalid sintaxis"<<'\n';
```
syntax

DimohaZadira replied:
`std::cout << "Invalid syntax" << std::endl;` вот так надо)

DimohaZadira replied:
ну либо `using std::cout` и тогда просто `cout`

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+﻿#include <iostream>
+#include <string>
+#include <cmath>
+using namespace std;
```
бан

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,39 @@
+            cout << "invalid sintaxis"<<'\n';
+        }
+    }
+    cout << ans;
```
`std::cout << ans << std::endl;`

smidy2002 committed:
поправил замечания из коментов

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,42 @@
+    uint32_t ans = 0;
+    const string maxValue = std::numeric_limits<uint32_t>.max();
+    if (n[0] == '0') { return -1; }
+    if (string_num[i] > '9' || string_num[i] < 48) { return -1; }
```
ну а 48 чем лучше?)

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,42 @@
+int main()
+{
+    string string_num = "-1";
+    int ans = 0;
```
почему остался int?

DimohaZadira commented on hw3/main.cpp:
```diff
@@ -0,0 +1,42 @@
+{
+    uint32_t i = 0;
+    uint32_t ans = 0;
+    const string maxValue = std::numeric_limits<uint32_t>.max();
```
ты запускал программу, она работает? Как минимум должно поругаться здесь

smidy2002 replied:
вроде запускается)

DimohaZadira replied:
кароч тут под капотом происходит неявное преобразование в строку из uint32_t, что кринж. Напиши явно

smidy2002 committed:
слишком много вывел из тела цикла, исправил

smidy2002 committed:
заменил инт, 48-> '0'

smidy2002 committed:
явное преобразование

std::numeric_limits<uint32_t>.max()      ->        std::to_string(std::numeric_limits<uint32_t>.max());
